### PR TITLE
Spring WebClient support

### LINKED ISSUE #1306: add support for reactive Spring WebClient instrumentation
## Is your feature request related to a problem?
[Webflux](https://docs.spring.io/spring/docs/current/spring-framework-reference/web-reactive.html) provides a client that allows to communicate with other reactive services.

While server-side instrumentation is covered with elastic/apm-agent-java#1305 , the client part is not part of it.

## Describe the solution you'd like
- instrument Webflux client
- capture external HTTP requests as spans and link them to the active transaction.
- should support the two available implementations: reactor/netty and jetty
- add context-propagation headers to HTTP calls

### Implementation notes
- instrument all sub-classes of `ClientHttpConnector` seems a good start

### PR DESCRIPTION
**Pull request by kananindzya**
## What does this PR do?

Fixes #1306 
continue from https://github.com/elastic/apm-agent-java/pull/1932

**NOTE:**
this PR added support for Spring WebClient only. But `spring-cloud-reactor-gateway` under the hood uses `reactor-netty-client`.

## Checklist

- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by kananindzya** [comment]
image
span details - here https://gist.github.com/kananindzya/06a4594e519c7ee6332faea990ad6a84

File: apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchHelper.java [hunk 1]
```diff
@@ -1,3 +1,21 @@
+/*
```
Comment (fern-we) [inline apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchHelper.java:1]: It feels like all of these rabbitmq changes should move to different PR?

Comment (kananindzya) [reply apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchHelper.java:1]: No need, this about missed licensed headers. This headers is already at master branch now.

**Comment by fern-we** [comment]
👍  very relevant!

**Comment by fern-we** [comment]
Lots of great progress. Any updates?

**Comment by alogishetty** [comment]
Hey, any updates on this PR? We are highly looking forward for this feature!

**Review by msalaslo — COMMENTED** [review]
reviwed

**Review by msalaslo — APPROVED** [review]
ok

**Comment by fern-we** [comment]
Updates? Are we close to merging? :)

**Comment by SylvainJuge** [comment]
Hi,

We plan do deal with this within short time-frame (hopefully only a few days/weeks), but I can't provide any ETA.
It has been close to the top of the list of our priorities for a while, but still not yet on the top of it, but we are working on it.

**Comment by gargrag** [comment]
Super Relevant!

**Comment by gargrag** [comment]
@SylvainJuge This is blocking us hard, to the point we're considering diverting to another tracing solution. Is there anything our teams can do to help expedite this improvement? Thanks!

**Comment by SylvainJuge** [comment]
Hi @gargrag , I am working on it this week, thus it should be a matter of a few days if nothing more urgent happens in the mean time (which tends to happen quite often unfortunately). Would you be interested in testing a snapshot to make sure that it works as expected in your use-case ?

**Comment by SylvainJuge** [comment]
Hi @kananindzya , I don't know why but the CLA check seems to not pass on some of your commits, could you open https://www.elastic.co/contributor-agreement and re-sign the CLA ? It seems that a different email address was used on the commits.

**Comment by kananindzya** [comment]
> Hi @kananindzya , I don't know why but the CLA check seems to not pass on some of your commits, could you open https://www.elastic.co/contributor-agreement and re-sign the CLA ? It seems that a different email address was used on the commits.

Hi @SylvainJuge please check again

**Comment by SylvainJuge** [comment]
@elasticmachine run elasticsearch-ci/docs

**Comment by SylvainJuge** [comment]
I only had to perform very minimal/cosmetic changes, thanks a lot @kananindzya and @abdulazizali77 for the work on this PR, this is awesome !

FYI, I tried to refactor the 3 implementations of `CoreSubscriber` to a simpler one but that proved to tedious, added an extra dependency to reactor plugin and meant revisiting/fixing most Webflux and Webclient tests which is definitely too much effort to make it worth for now, maybe we'll revisit that when we instrument another component based on reactor or reactive streams.

**Comment by SylvainJuge** [comment]
@elasticmachine run elasticsearch-ci/docs

**Comment by abdulazizali77** [comment]
Congrats for pushing this through @SylvainJuge. Apologies for airing dirty laundry here, notwithstanding how badly designed, badly implemented, badly timed, or badly intended my own contribution and work is; suffice to say due to the pushiness and entitlement of @fern-we roping in @kananindzya its unfortunate i never want to contribute to elastic/apm-agent ever again.

**Comment by SylvainJuge** [comment]
> Congrats for pushing this through @SylvainJuge. Apologies for airing dirty laundry here, notwithstanding how badly designed, badly implemented, badly timed, or badly intended my own contribution and work is; suffice to say due to the pushiness and entitlement of @fern-we roping in @kananindzya its unfortunate i never want to contribute to elastic/apm-agent ever again.

I am really sorry that you've felt offended by what happened here @abdulazizali77 .

You initiated this plugin in https://github.com/elastic/apm-agent-java/pull/1932 PR and your commits served as the basis of this PR as we can see in the [commit log](https://github.com/elastic/apm-agent-java/pull/2229/commits), so you still deserve most of the credit for this.

What @kananindzya did (or what I assume the intent is) was just trying to move this forward through another PR, because he is not an Elastic employee he could not directly push nor directly contribute to your PR, if we had to do it again I would have suggested him opening a PR against your own fork/branch instead of opening another one based on your work.

There is nothing wrong with your contribution here, even if the code you contributed was later removed or modified, it helped building the current form which we are about to ship and we should all feel grateful for that. Unfortunately, working on agents means that we often have to iterate a lot and often rewrite code often more than once with different approaches before being able to ship something.
